File: LeetCode/Medium/Letter Combinations Of A Phone Number/Go/letterCombinationOfPhoneNumber.go
package main

import "fmt"

var numberLetterMap = map[byte]string{
	'1': "", '2': "abc", '3': "def",
	'4': "ghi", '5': "jkl", '6': "mno",
	'7': "pqrs", '8': "tuv", '9': "wxyz",
}

func letterCombinations(digits string) []string {
	res := []string{}
	backtrack(digits, 0, "", &res)
	return res
}

func backtrack(digits string, ind int, temp string, res *[]string) {
	if ind == len(digits) {
		if temp != "" {
			*res = append(*res, temp)
		}
		return
	}
	letters := numberLetterMap[digits[ind]]
	for i := 0; i < len(letters); i++ {
		temp += string(letters[i])
		backtrack(digits, ind+1, temp, res)
		temp = temp[:len(temp)-1]
	}
}

func main() {
	fmt.Println(letterCombinations("23"))
}

File: LeetCode/Medium/Letter Combinations Of A Phone Number/C++/letter_combination_of_phone_number.cc
typedef vector<string> vs;

class Solution {
public:
    map<char, string> digit_char;
    vector<string> letterCombinations(string digits) {
        vs res;
        digit_char = {{'1', ""}, {'2', "abc"},
                    {'3', "def"}, {'4', "ghi"},
                    {'5', "jkl"}, {'6', "mno"},
                    {'7', "pqrs"}, {'8', "tuv"},
                    {'9', "wxyz"}};
        backtrack(digits, 0, res, "");
        return res;
    }

    void backtrack(string digits, int ind, vs& res, string temp) {
        if (ind == digits.length()) {
            if (temp.length() > 0) res.push_back(temp);
            return;
        }

        string st = digit_char.at(digits[ind]);
        for (int i=0; i<st.length(); i++) {
            temp += st[i];
            backtrack(digits, ind+1, res, temp);
            temp.pop_back();
        }
    }
};
